### PR TITLE
[FLINK-31677][table] Add built-in MAP_ENTRIES function.

### PR DESCRIPTION
**Pull request by liuyongvs**
- What is the purpose of the change
This is an implementation of MAP_ENTRIES

- Brief change log
MAP_ENTRIES for Table API and SQL
    
```
map_entries(map) - Returns an unordered array of all entries in the given map.
Syntax:
    map_entries(map)
Arguments:
    map: An MAP to be handled.
Returns:Returns an unordered array of all entries in the given map. Returns null if the argument is null

> SELECT map_entries(map[1, 'a', 2, 'b']);
 [(1,"a"),(2,"b")] 
```

See also
presto https://prestodb.io/docs/current/functions/map.html
spark https://spark.apache.org/docs/latest/api/sql/index.html#map_entries

- Verifying this change
This change added tests in MapFunctionITCase.

- Does this pull request potentially affect one of the following parts:
Dependencies (does it add or upgrade a dependency): ( no)
The public API, i.e., is any changed class annotated with @Public(Evolving): (yes )
The serializers: (no)
The runtime per-record code paths (performance sensitive): ( no)
Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
The S3 file system connector: ( no)
- Documentation
Does this pull request introduce a new feature? (yes)
If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by liuyongvs** [comment]
hi @snuyanzin do you have time to review it?

**Review by Aitozi — APPROVED** [review]
LGTM

**Comment by liuyongvs** [comment]
hi @dawidwys could you have a look this ,it is long time no one to review and merge?

File: docs/data/sql_functions.yml [hunk 1]
```diff
@@ -646,6 +646,9 @@ collection:
   - sql: MAP_VALUES(map)
     table: MAP.mapValues()
     description: Returns the values of the map as array. No order guaranteed.
+  - sql: MAP_ENTRIES(map)
+    table: MAP.mapEntries()
+    description: Returns an unordered array of all entries in the given map. No order guaranteed.
```
Comment (snuyanzin) [inline docs/data/sql_functions.yml:651]: what means `unordered` here?

File: flink-python/docs/reference/pyflink.table/expressions.rst [hunk 2]
```diff
@@ -231,6 +231,7 @@ advanced type helper functions
     Expression.array_remove
     Expression.map_keys
     Expression.map_values
+    Expression.map_entries
```
Comment (snuyanzin) [inline flink-python/docs/reference/pyflink.table/expressions.rst:234]: it's better to have abc order

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 3]
```diff
@@ -1396,6 +1397,11 @@ public OutType mapValues() {
         return toApiSpecificExpression(unresolvedCall(MAP_VALUES, toExpr()));
     }
 
+    /** Returns an unordered array of all entries in the given map. */
```
Comment (snuyanzin) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:1400]: same about `unordered`

File: docs/data/sql_functions.yml (hunk 1 above)
Comment (liuyongvs) [reply docs/data/sql_functions.yml:651]: keys are not sorted, i will remove it @snuyanzin

**Comment by liuyongvs** [comment]
rebase and fix your reviews @snuyanzin

**Comment by liuyongvs** [comment]
hi @snuyanzin ci passed

**Comment by liuyongvs** [comment]
hi @snuyanzin do you have time look again?

**Comment by liuyongvs** [comment]
hi @snuyanzin

**Comment by liuyongvs** [comment]
hi @snuyanzin do you have time look again?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificTypeStrategies.java [hunk 4]
```diff
@@ -89,6 +95,49 @@ public final class SpecificTypeStrategies {
...
+                            DataTypes.ARRAY(
+                                    ((KeyValueDataType) callContext.getArgumentDataTypes().get(0))
+                                            .getValueDataType()));
+
+    /** Type strategy specific for {@link BuiltInFunctionDefinitions#MAP_ENTRIES}. */
+    public static final TypeStrategy MAP_ENTRIES =
+            callContext ->
+                    Optional.of(
+                            DataTypes.ARRAY(
+                                    DataTypes.ROW(
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificTypeStrategies.java:119]: Do we need to specify the name for  each field of the `Row` type? Btw, seems spark use
```
      StructType(
        StructField("key", childDataType.keyType, false) ::
        StructField("value", childDataType.valueType, childDataType.valueContainsNull) ::
        Nil)
````

**Review by luoyuxia — COMMENTED** [review]
@liuyongvs Thanks for this PR. I left one comment.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificTypeStrategies.java (hunk 4 above)
Comment (liuyongvs) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificTypeStrategies.java:119]: added @luoyuxia

**Pull request merged**